File: src-tauri/src/commands/charts.rs
use tauri::State;

use notecli::models::{
    ActiveUsersChart, ApRequestChart, FederationChart, ServerDriveChart, ServerNotesChart,
    ServerUsersChart, UserFollowingChart, UserNotesChart, UserPvChart,
};

use super::{get_credentials_or_anon, AppState, Result};

// チャート系エンドポイントは public (未ログインでも閲覧可)。

#[tauri::command]
#[specta::specta]
pub async fn api_charts_user_notes(
    app_state: State<'_, AppState>,
    account_id: String,
    params: serde_json::Value,
) -> Result<UserNotesChart> {
    let (db, client) = app_state.ready().await;
    let (host, token) = get_credentials_or_anon(&db, &account_id)?;
    let raw = client
        .request(&host, &token, "charts/user/notes", params)
        .await?;
    Ok(serde_json::from_value(raw)?)
}

#[tauri::command]
#[specta::specta]
pub async fn api_charts_user_following(
    app_state: State<'_, AppState>,
    account_id: String,
    params: serde_json::Value,
) -> Result<UserFollowingChart> {
    let (db, client) = app_state.ready().await;
    let (host, token) = get_credentials_or_anon(&db, &account_id)?;
    let raw = client
        .request(&host, &token, "charts/user/following", params)
        .await?;
    Ok(serde_json::from_value(raw)?)
}

#[tauri::command]
#[specta::specta]
pub async fn api_charts_user_pv(
    app_state: State<'_, AppState>,
    account_id: String,
    params: serde_json::Value,
) -> Result<UserPvChart> {
    let (db, client) = app_state.ready().await;
    let (host, token) = get_credentials_or_anon(&db, &account_id)?;
    let raw = client
        .request(&host, &token, "charts/user/pv", params)
        .await?;
    Ok(serde_json::from_value(raw)?)
}

#[tauri::command]
#[specta::specta]
pub async fn api_charts_active_users(
    app_state: State<'_, AppState>,
    account_id: String,
    params: serde_json::Value,
) -> Result<ActiveUsersChart> {
    let (db, client) = app_state.ready().await;
    let (host, token) = get_credentials_or_anon(&db, &account_id)?;
    let raw = client
        .request(&host, &token, "charts/active-users", params)
        .await?;
    Ok(serde_json::from_value(raw)?)
}

#[tauri::command]
#[specta::specta]
pub async fn api_charts_notes(
    app_state: State<'_, AppState>,
    account_id: String,
    params: serde_json::Value,
) -> Result<ServerNotesChart> {
    let (db, client) = app_state.ready().await;
    let (host, token) = get_credentials_or_anon(&db, &account_id)?;
    let raw = client
        .request(&host, &token, "charts/notes", params)
        .await?;
    Ok(serde_json::from_value(raw)?)
}

#[tauri::command]
#[specta::specta]
pub async fn api_charts_users(
    app_state: State<'_, AppState>,
    account_id: String,
    params: serde_json::Value,
) -> Result<ServerUsersChart> {
    let (db, client) = app_state.ready().await;
    let (host, token) = get_credentials_or_anon(&db, &account_id)?;
    let raw = client
        .request(&host, &token, "charts/users", params)
        .await?;
    Ok(serde_json::from_value(raw)?)
}

#[tauri::command]
#[specta::specta]
pub async fn api_charts_federation(
    app_state: State<'_, AppState>,
    account_id: String,
    params: serde_json::Value,
) -> Result<FederationChart> {
    let (db, client) = app_state.ready().await;
    let (host, token) = get_credentials_or_anon(&db, &account_id)?;
    let raw = client
        .request(&host, &token, "charts/federation", params)
        .await?;
    Ok(serde_json::from_value(raw)?)
}

#[tauri::command]
#[specta::specta]
pub async fn api_charts_ap_request(
    app_state: State<'_, AppState>,
    account_id: String,
    params: serde_json::Value,
) -> Result<ApRequestChart> {
    let (db, client) = app_state.ready().await;
    let (host, token) = get_credentials_or_anon(&db, &account_id)?;
    let raw = client
        .request(&host, &token, "charts/ap-request", params)
        .await?;
    Ok(serde_json::from_value(raw)?)
}

#[tauri::command]
#[specta::specta]
pub async fn api_charts_drive(
    app_state: State<'_, AppState>,
    account_id: String,
    params: serde_json::Value,
) -> Result<ServerDriveChart> {
    let (db, client) = app_state.ready().await;
    let (host, token) = get_credentials_or_anon(&db, &account_id)?;
    let raw = client
        .request(&host, &token, "charts/drive", params)
        .await?;
    Ok(serde_json::from_value(raw)?)
}

File: src/adapters/misskey/api/charts.ts
import { commands, unwrap } from '@/utils/tauriInvoke'
import type {
  ActiveUsersChart,
  ApRequestChart,
  ChartsApi,
  FederationChart,
  ServerDriveChart,
  ServerNotesChart,
  ServerUsersChart,
  UserFollowingChart,
  UserNotesChart,
  UserPvChart,
} from '../../types'
import type { MisskeyApiContext } from './context'

// chart 型は specta 生成の bindings と adapters/types.ts の re-export が一致
// するため、他ドメインと違い unwrapAny を要さずそのまま unwrap できる。
export function createChartsApi(ctx: MisskeyApiContext): ChartsApi {
  return {
    async getUserNotesChart(
      userId: string,
      span: 'day' | 'hour' = 'day',
      limit = 350,
    ): Promise<UserNotesChart> {
      return unwrap(
        await commands.apiChartsUserNotes(ctx.accountId, {
          userId,
          span,
          limit,
        }),
      )
    },

    async getUserFollowingChart(
      userId: string,
      span: 'day' | 'hour' = 'day',
      limit = 30,
    ): Promise<UserFollowingChart> {
      return unwrap(
        await commands.apiChartsUserFollowing(ctx.accountId, {
          userId,
          span,
          limit,
        }),
      )
    },

    async getUserPvChart(
      userId: string,
      span: 'day' | 'hour' = 'day',
      limit = 30,
    ): Promise<UserPvChart> {
      return unwrap(
        await commands.apiChartsUserPv(ctx.accountId, { userId, span, limit }),
      )
    },

    async getActiveUsersChart(
      span: 'day' | 'hour' = 'day',
      limit = 90,
    ): Promise<ActiveUsersChart> {
      return unwrap(
        await commands.apiChartsActiveUsers(ctx.accountId, { span, limit }),
      )
    },

    async getServerNotesChart(
      span: 'day' | 'hour' = 'day',
      limit = 90,
    ): Promise<ServerNotesChart> {
      return unwrap(
        await commands.apiChartsNotes(ctx.accountId, { span, limit }),
      )
    },

    async getServerUsersChart(
      span: 'day' | 'hour' = 'day',
      limit = 90,
    ): Promise<ServerUsersChart> {
      return unwrap(
        await commands.apiChartsUsers(ctx.accountId, { span, limit }),
      )
    },

    async getFederationChart(
      span: 'day' | 'hour' = 'day',
      limit = 90,
    ): Promise<FederationChart> {
      return unwrap(
        await commands.apiChartsFederation(ctx.accountId, { span, limit }),
      )
    },

    async getApRequestChart(
      span: 'day' | 'hour' = 'day',
      limit = 90,
    ): Promise<ApRequestChart> {
      return unwrap(
        await commands.apiChartsApRequest(ctx.accountId, { span, limit }),
      )
    },

    async getServerDriveChart(
      span: 'day' | 'hour' = 'day',
      limit = 90,
    ): Promise<ServerDriveChart> {
      return unwrap(
        await commands.apiChartsDrive(ctx.accountId, { span, limit }),
      )
    },
  }
}
